### ISSUE TITLE
Add 'baseName' option and remove 'basenameSeparator' option of TranslationCheck

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#2204]
Based on https://github.com/checkstyle/checkstyle/pull/2177 and discussion https://github.com/MEZk/checkstyle/commit/4b282073d8b37fc11881f5f1e834f55ca33ed2c0#commitcomment-13272046 Translation check needs new option which allows to specify resource bundles names which should be skipped while validating existence of translations. Default value should be empty String Set which means that all resource bundles should be validated.

**Comment by romani** [comment]
#2251 issue need to be resolved first to make it clear what is logic of basename and language/country/variant processing.

**Comment by romani** [comment]
this issue need to be changed to use positive and exact logic.
Property files are used not only for Locale/Translation, in most cases they store configurations.
Instead of ignore list we might have RegExp  field that restrict validation on certain set of files.

**Comment by MEZk** [comment]
1) I'm for new option "baseName" of RegExp type. 
2) Agree.

3) 
3.1) 
```
Resource bundle 'ButtonLabel':
    |---ButtonLabel.properties
    |---ButtonLabel_de.properties
```
```ResourceBundle.getBundle("ButtonLabel", new Locale("de"));``` returns  bundle for German locale.

3.2) 
The following files does not form a resource bundle:
```
ButtonLabel.properties
ButtonLabel-de.properties
```
```ResourceBundle.getBundle("ButtonLabel", new Locale("de"));``` returns bundle for default locale, since ```ButtonLabel_de.properties``` is not found. Resource bundle with base name 'ButtonLabel' contains only one resource file for default locale.

3.3) 
```
ButtonLabel-de.properties
```

```ResourceBundle.getBundle("ButtonLabel", new Locale("de"));``` throws an exception: 
```Method threw 'java.util.MissingResourceException' exception.```

3.4) 
```
ButtonLabel-de.properties
```

```ResourceBundle.getBundle("ButtonLabel-de", new Locale("de"));``` returns the resource bundle with the base name 'ButtonLabel-de' for default locale. It does not mean that 'ButtonLabel-de' is a resource bundle for German locale!

4) Once again, my proposals are:
* Remove the 'basenameSeparator' option.
* Add new option "baseName" of RegExp type. If user does not specify the option we should parse base name as the following:

   a) messages.properties

   ```java
   String baseName = "messages.properties".substring(0, "messages.properties".lastIndexOf('.'));
   ```
  b) If resource file for default locale (messages.properties) does not exist and there is only resource file for specific language (it might be messages_de.properties, messages_de_GE.properties, messages_de_GE_UNIX.properties or even messages_home_de.properties) we should somehow substring all characters before _de as a base name. Maybe we can use some regexp masks, forExample:

````java
public class TestExtraxtBaseName {
    public String getBaseName(String bundleName) {
        final String languageCountryVariant = "^.*_[a-z]{2}_[A-Z]{2}_[A-Za-z]*.properties$";
        final String languageCountry = "^.*_[a-z]{2}_[A-Z]{2}.properties$";
        final String language = "^.*_[a-z]{2}.properties$";
        
        final String removePattern;
        if (Pattern.matches(languageCountryVariant, bundleName)) {
            removePattern = "_[a-z]{2}_[A-Z]{2}_[A-Za-z]*.properties$";
        }
        else if (Pattern.matches(languageCountry, bundleName)) {
            removePattern = "_[a-z]{2}_[A-Z]{2}.properties$";
        }
        else if (Pattern.matches(language, bundleName)) {
            removePattern = "_[a-z]{2}.properties$";
        }
        else {
            // default 
            removePattern = ".properties$";
        }
        return StringUtils.removePattern(bundleName, removePattern);
    }
}
````

5) We can improve the check and validate the language, country, variant code if requiredTranslations set is not empty.

See suggestions at 
http://stackoverflow.com/questions/3684747/how-to-validate-a-locale-in-java

**Comment by romani** [comment]
@MEZk ,

> Remove the 'basenameSeparator' option.

I am ok too.

> Add new option "baseName" of RegExp type

basename or baseName ? 

> Add new option "baseName" of RegExp type. If user does not specify the option we should parse base name as the following:

we can not distinguish property file with translations and with configurations. Lets do nothing by default, as we can not reliably say what is a target of this Check.

> 5) We can improve the check and validate the language, country, variant code if requiredTranslations set is not empty.

please split this to separate issue.

> 6) In addition, we should check not only files with '.resources' extension, but also all files which extensions are specified for the check!

I think you mean "*.properties" by "*.resources". Yes Check need to check any files that user want. But please beware of `extension` option of Checker.

**Comment by MEZk** [comment]
>basename or baseName ?

I think 'baseName', because [here](https://docs.oracle.com/javase/tutorial/i18n/resbundle/concept.html) they use two seperate worlds: base and name. We can use Camel case for the name of the option.

**Comment by romani** [comment]
ok. let do `baseName`, java use CaleCase https://docs.oracle.com/javase/7/docs/api/java/util/ResourceBundle.html#getBundle%28java.lang.String,%20java.util.Locale%29

**Comment by MEZk** [comment]
@romani 
>we can not distinguish property file with translations and with configurations. Lets do nothing by default, as we can not reliably say what is a target of this Check.

The check arragnes bundles as a map where key is a baseName. That is why we should substring baseName and then find all files which are included in resource bundle. If user do not specify baseName we can not collect files and organize the map.

**Comment by romani** [comment]
"property" files are used for multiple purposes. In most cases it is not related to localizations. Localization is so rare in software. Localization is too complicated and hard to maintain, so only big project care about it.

can we reliably say that if name contains language code it is a localization property file ?
I think - no. it is "yes" in most cases, but not in all. 

For case:
```
1) messages
2) messages_buttons
3) messages_home
4) messages_home_excluded
```

regexp could be `messages*` . It will help us to trigger Check there , and after this use smart logic to parse out language+variant and get base name. User need to help us to distinguish config file and localization file.

**Comment by MEZk** [comment]
@romani 
Ok, now I understand what you mean.

**Comment by MEZk** [comment]
@romani 
From ResourceBundle#getBundle documentation:
>getBundle attempts to locate a property resource file using the generated properties file name. It generates a path name from the candidate bundle name by replacing all "." characters with "/" and appending the string ".properties". It attempts to find a "resource" with this name using ClassLoader.getResource. (Note that a "resource" in the sense of getResource has nothing to do with the contents of a resource bundle, it is just a container of data, such as a file.) If it finds a "resource", it attempts to create a new PropertyResourceBundle instance from its contents. If successful, this instance becomes the result resource bundle. 

Looks like we should check only files which have '.properties' extension.

**Comment by romani** [comment]
Yes, but not all properties file are our target for translation validation.

**Comment by romani** [comment]
fix is merged.
